### ISSUE TITLE
Packed varint encoding is not forwards-compatible for field numbers greater than 2047 in C++

### ISSUE THREAD

**Issue by grandseiken** [google/protobuf#2879]
As stated in the [encoding documentation](https://developers.google.com/protocol-buffers/docs/encoding#optional),

> Protocol buffer parsers must be able to parse repeated fields that were compiled as packed as if they were not packed, and vice versa. This permits adding [packed=true] to existing fields in a forward- and backward-compatible way.

This does not seem to be true for the 3.1.0 C++ protobuf parser. Generated code for a repeated field with `[packed=false]` cannot parse a message serialized by generated code with `[packed=true]` if the field number is greater than 2047. (Incidentally, this is the largest field number for which the varint-encoded (field number, wire type) key is at most 2 bytes long.)

Reproduction case:

```
// test.proto
syntax = "proto2";
package test;

message Packed {
  repeated int32 a = 2048 [packed=true];
}

message NotPacked {
  repeated int32 a = 2048;
}
```

```
// test.cc
#include "test.pb.h"
#include <gtest/gtest.h>
#include <string>

TEST(Test, Packed2048) {
  test::Packed packed;
  packed.add_a(1234);
  std::string buffer;
  packed.SerializeToString(&buffer);
  
  test::NotPacked not_packed;
  not_packed.ParseFromString(buffer);
  EXPECT_EQ(1, not_packed.a_size()); // Actual: 0
}
```

The test passes if any other combination of packed/non-packed encodings is used for the two messages, or if the field number is changed to 2047 or less.

Update: after a little more testing, this only seems to apply to varint-encoded packed fields, i.e. changing the field types to `fixed32` above also makes the test pass. Also, I'm using 3.1.0, not 3.0.1 as I erroneously stated.

**Comment by acozzette** [comment]
@grandseiken Thank you for reporting this. I was able to reproduce it on my machine as well. Here is the problematic parsing code generated by `protoc`:

```
  for (;;) {
    ::std::pair< ::google::protobuf::uint32, bool> p = input->ReadTagWithCutoffNoLastTag(16384u);
    tag = p.first;
    if (!p.second) goto handle_unusual;
    switch (::google::protobuf::internal::WireFormatLite::GetTagFieldNumber(tag)) {
      // repeated int32 a = 2048;
      case 2048: {
        if (static_cast< ::google::protobuf::uint8>(tag) ==
            static_cast< ::google::protobuf::uint8>(16384u)) {
          DO_((::google::protobuf::internal::WireFormatLite::ReadRepeatedPrimitive<
                   ::google::protobuf::int32, ::google::protobuf::internal::WireFormatLite::TYPE_INT32>(
                 3, 16384u, input, this->mutable_a())));
        } else if (static_cast< ::google::protobuf::uint8>(tag) ==
                   static_cast< ::google::protobuf::uint8>(16386u)) {
          DO_((::google::protobuf::internal::WireFormatLite::ReadPackedPrimitiveNoInline<
                   ::google::protobuf::int32, ::google::protobuf::internal::WireFormatLite::TYPE_INT32>(
                 input, this->mutable_a())));
        } else {
          goto handle_unusual;
        }
        break;
      }

      default: {
      handle_unusual:
        if (tag == 0 ||
            ::google::protobuf::internal::WireFormatLite::GetTagWireType(tag) ==
            ::google::protobuf::internal::WireFormatLite::WIRETYPE_END_GROUP) {
          goto success;
        }
        DO_(::google::protobuf::internal::WireFormatLite::SkipField(
            input, tag, &unknown_fields_stream));
        break;
      }
    }
  }
```
I think the problem occurs when we call `input->ReadTagWithCutoffNoLastTag(16384u)`; we're passing too low a cutoff value there because we're not taking into account the possibility of a higher tag due to the packed field having a wire type of 2 instead of 0. This results in the field being treated as an unknown field.

**Comment by acozzette** [comment]
I forgot to close this issue earlier but it was fixed in 0400cca3236de1ca303af38bf81eab332d042b7c.
